### PR TITLE
Add initial unit tests for slurmdbd charm

### PR DESCRIPTION
**Pull request by NucciTheBoss**
## Description

This pull request adds an initial unit test suite for the slurmdbd charm. It aims to cover most of the methods defined in `charm.py` and "boundary cases" (i.e. is slurmdbd failing to install properly handled). The idea is to get feedback on these unit tests to help guide decisions for future slurmdbd charm unit tests. 

## How was the code tested?

I ran `tox -e unit` locally on my system. __Note:__ I intentionally have some unit tests failing. The tests are failing due to missing files that would have been created by Omni's old make file.

## Related issues and/or tasks

This pull request is part of the Charmed SLURM project on GitHub.

## Checklist

- [x] I am the author of these changes, or I have the rights to submit them.
- [x] I have added the relevant changes to the README and/or documentation.
- [x] I have self reviewed my own code.
- [x] All requested changes and/or review comments have been resolved.

### CODE REVIEW

File: tests/unit/test_charm.py [hunk 1]
```diff
@@ -1,25 +1,147 @@
...
+        install.side_effect = False
+        self.harness.charm.on.install.emit()
+        self.assertEqual(
+            self.harness.charm.unit.status, BlockedStatus("Error installing slurmdbd")
+        )
+
+    @patch("pathlib.Path.read_text", return_value="v1.0.0")
+    def test_upgrade_charm(self, _) -> None:
+        """Test that charm upgrade procedure works."""
+        self.harness.charm.on.upgrade_charm.emit()
```
Comment (dvdgomez) [inline tests/unit/test_charm.py:51]: Missing assert?

Comment (NucciTheBoss) [reply tests/unit/test_charm.py:51]: No, this is intentional. I am trying to catch if there is an error thrown when ops updates the version of the charm. Looking through the documentation, I did not see any property that returns the version of the charm. I could probably add a patched defer object that asserts that the method was called.

File: tests/unit/test_charm.py [hunk 2]
```diff
@@ -1,25 +1,147 @@
...
+
+    @patch("ops.model.Unit.is_leader", return_value=True)
+    def test_is_leader(self, _) -> None:
+        """Test that _is_leader method works.
+
+        Notes:
+            The _is_leader method should be removed from charm.py
+            since _is_leader is already defined in ops.
+        """
+        self.assertEqual(self.harness.charm._is_leader(), True)
```
Comment (dvdgomez) [inline tests/unit/test_charm.py]: I think it's more clear to change this to `self.assertTrue(self.harness.charm._is_leader())`. Same thing with other cases below to `self.assertFalse` instead of `self.assertEqual(..., False)`.

File: tests/unit/test_charm.py [hunk 3]
```diff
@@ -1,25 +1,147 @@
...
+
+    def test_on_slurmctld_unavailable(self) -> None:
+        """Test that _on_slurmctld_unavailable method works."""
+        self.harness.charm.on.slurmctld_unavailable.emit()
+        self.assertEqual(self.harness.charm._stored.jwt_available, False)
+        self.assertEqual(self.harness.charm._stored.munge_available, False)
+
+    def test_write_config_and_restart_slurmdbd(self) -> None:
+        """Test that _write_config_and_restart_slurmdbd method works."""
+        self.harness.charm.on.write_config.emit()
```
Comment (dvdgomez) [inline tests/unit/test_charm.py]: Missing assert?

File: tests/unit/test_charm.py (hunk 1 above)
Comment (dvdgomez) [reply tests/unit/test_charm.py:51]: Oh gotcha, that makes sense. Could you add that in a comment or update the test description? At first glance it looks like something is missing.

File: tests/unit/test_charm.py (hunk 3 above)
Comment (dvdgomez) [reply tests/unit/test_charm.py]: If similar to the upgrade test case just add a comment or update test description about the test failing on catching an error.

File: tests/unit/test_charm.py (hunk 1 above)
Comment (NucciTheBoss) [reply tests/unit/test_charm.py:51]: Yes, that is a good idea. I can add a `Notes:` section to the docstring that explains that I am testing for if an exception is raised.

File: tests/unit/test_charm.py (hunk 3 above)
Comment (NucciTheBoss) [reply tests/unit/test_charm.py]: See https://github.com/canonical/slurmdbd-operator/pull/2#discussion_r1065966052

File: tests/unit/test_charm.py (hunk 1 above)
Comment (NucciTheBoss) [reply tests/unit/test_charm.py:51]: I reworked this method to where I am now using `assertEqual()` to compare workload versions instead of just invoking `emit()` in https://github.com/canonical/slurmdbd-operator/pull/2/commits/6ad37cd848f9a2dcbee3fb1df15b742d19f45260

File: tests/unit/test_charm.py (hunk 3 above)
Comment (NucciTheBoss) [reply tests/unit/test_charm.py]: Decided to bomb this unit test out in https://github.com/canonical/slurmdbd-operator/pull/2/commits/6ad37cd848f9a2dcbee3fb1df15b742d19f45260 since this is something that I am planning to change in the charm.

File: tests/unit/test_charm.py (hunk 2 above)
Comment (NucciTheBoss) [reply tests/unit/test_charm.py]: Resolved in https://github.com/canonical/slurmdbd-operator/pull/2/commits/6ad37cd848f9a2dcbee3fb1df15b742d19f45260. I kept assertEqual but compare the returned value of the locally defined `_is_leader()` method with the returned value of ops `is_leader()` method.

**Comment by dvdgomez** [comment]
If you have a test that fails currently and you expect it to then I would this add decorator `@unittest.expectedFailure`. After we refactor where it works then we can remove the decorator

**Comment by NucciTheBoss** [comment]
> If you have a test that fails currently and you expect it to then I would this add decorator `@unittest.expectedFailure`. After we refactor where it works then we can remove the decorator

Good idea. Let me implement this. Also need to figure out why property mocking is not working as I would like it too.

**Review by dvdgomez — APPROVED** [review]
Nicely done, good work!
